samcli/commands/init/init_templates.py

```diff
@@ -118,23 +120,72 @@ def _shared_dir_check(self, shared_dir):
+                LOG.info("\nCloning app templates from %s", self._repo_url)
                 subprocess.check_output(
-                    [self._git_executable(), "clone", self._repo_url], cwd=shared_dir, stderr=subprocess.STDOUT
+                    [self._git_executable(), "clone", self._repo_url, self._repo_name],
+                    cwd=tempdir,
+                    stderr=subprocess.STDOUT,
                 )
+                # Now we need to delete the old repo and move this one.
+                try:
+                    LOG.debug("Removing old templates from %s", str(expected_path))
+                    shutil.rmtree(expected_path)
+                    LOG.debug("Copying templates from %s to %s", str(expected_temp_path), str(expected_path))
+                    shutil.copytree(expected_temp_path, expected_path, ignore=shutil.ignore_patterns("*.git"))
+                except (OSError, shutil.Error) as ex:
+                    # UNSTABLE STATE - it's difficult to see how this scenario could happen except weird permissions, user will need to debug
+                    raise AppTemplateUpdateException(
+                        "Unstable state when updating app templates. Check that you have permissions to create/delete files in the AWS SAM shared directory or file an issue at https://github.com/awslabs/aws-sam-cli/issues"
+                    )
+                self.repo_path = expected_path
+            except OSError as ex:
+                LOG.warning("WARN: Could not clone app template repo.", exc_info=ex)
+            except subprocess.CalledProcessError as clone_error:
+                output = clone_error.output.decode("utf-8")
+                if "not found" in output.lower():
+                    click.echo("WARN: Could not clone app template repo.")
+
+    def _clone_new_app_templates(self, shared_dir, expected_path):
+        with osutils.mkdir_temp(ignore_errors=True) as tempdir:
+            expected_temp_path = os.path.normpath(os.path.join(tempdir, self._repo_name))
+            try:
+                LOG.info("\nCloning app templates from %s", self._repo_url)
+                subprocess.check_output(
+                    [self._git_executable(), "clone", self._repo_url], cwd=tempdir, stderr=subprocess.STDOUT
+                )
+                shutil.copytree(expected_temp_path, expected_path, ignore=shutil.ignore_patterns("*.git"))
                 self.repo_path = expected_path
             except OSError as ex:
                 LOG.warning("WARN: Can't clone app repo, git executable not found", exc_info=ex)
             except subprocess.CalledProcessError as clone_error:
                 output = clone_error.output.decode("utf-8")
                 if "not found" in output.lower():
                     click.echo("WARN: Could not clone app template repo.")
-        self.clone_attempted = True
+
+    def _template_directory_exists(self, expected_path):
+        path = Path(expected_path)
```

Comment:
nit/non-blocking: I notice we are using both pathlib and os.path, is it possible to standardize on one or other.